runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -502,6 +690,8 @@ public void close() throws Exception {
     processWatermark1(Watermark.MAX_WATERMARK);
     while (getCurrentOutputWatermark() < Watermark.MAX_WATERMARK.getTimestamp()) {
       invokeFinishBundle();
+      // Sleep for 5s to wait for any timer to be fired.
+      Thread.sleep(5000);
```

Comment:
This shouldn't be necessary. All timers will be drained on close(), at least in super.close().

Reply:
From my testing, I found that if we don't free the occupation here, the `DoFnOperator.onProcessingTime` will never get a chance to be executed.

Reply:
Kindly pinged : )

Reply:
This is a bug then. All processing timers should be fired on `close()`. It would be great to get rid of the sleep which can have unexpected side effects, e.g. when operators are chained in Flink.

Reply:
This was not a bug before we support sdf initiate checkpoint because normal processing time timer doesn't hold watermark. Now we are setting processing time timer to reschedule sdf residuals which holds the watermark back. When entering `ExecutableStageDoFnOperator.close()`, we will always in the while loop until all sdf processing timers get fired and watermark holds are removed. So the comment ` // Sleep for 5s to wait for any timer to be fired.` is not precise. The comment should be ` // Sleep for 5s to wait for any SDF timer to be fired.`

Reply:
Would you like to explain more about the side effect here?

Reply:
Makes sense. Thanks for clarifying. Is there a way to only sleep if there are such pending timers and then wait for precisely as long as the maximum of such timers? It just seems like a workaround to always wait 5 seconds. Also, I'm surprised these timers get fired at all, since newer versions of Flink stop processing timer execution once closing the operator (that's why we are draining the timers manually in `super.close()`).

Reply:
>  Is there a way to only sleep if there are such pending timers and then wait for precisely as long as the maximum of such timers?
One way I'm coming up with is we only sleep when we are executing SDF/Process fn.

> Also, I'm surprised these timers get fired at all, since newer versions of Flink stop processing timer execution once closing the operator (that's why we are draining the timers manually in super.close()).
I haven't looked into Flink timer service yet but I guess it might be related to watermark hold.

Reply:
Updated the logic in https://github.com/apache/beam/pull/13105/commits/d8bb778e64fc591819a90677a9f2e119e706b52f

Reply:
Thanks. This still feels like a workaround but at least we have it isolated for SDF only.

Reply:
I found that `Thread.sleep(5s)` is not a correct solution. The reason why it worked occasionally  is that the bundle is small enough and I got luck on race condition. I think we should figure out a way to be able to fire the SDF processing time timer when a bundle is closed within the life cycle of one `ExecutableStageDoFnOperator`.
Please correct me if I'm understanding it wrong:
* Flink starts all operators at the same time and closes the operators when the input watermark reaches MAX_TIMESTAMP, or it closes operators in a reverse topological order and `close()` is a blocking call?
* The processing time timers will not be fired anymore by the system once the `operator.close()` is invoked.
* The assumption around `ExecutableStageDoFnOperator` is that there is only one bundle executing inside one operator. When the output watermark advances to MAX_TIMESTAMP, we consider this bundle completed.

With supporting SDF initiated checkpoint, we do need to have several bundles invoked inside one `ExecutableStageDoFnOperator` life cycle, which means we either:

* Enable Flink to fire processing time timers after `Operator.close()` is invoked -- this may not be preferrable.
* Or we try to close the bundle before we reach to the `Operator.close()`.
* Or we manually drain SDF timers with scarifying the ability of `resumeDelay()`. For example, the user may want to reschedule the SDF residuals in 5 mins but we have to fire it now.

Do you have any ideas/suggestions? Thanks for your help!

Reply:
Double check the implementation of `DoFnOperator` and `ExecutableStageDoFnOperator`, we already invokes `finishBundle` when reaching 1000 input elements or 1s processing time by default. 
The real problem for SDF is that it's natural for SDF to read from `Impluse` and execute as a high fan-out DoFn. Based on current structure, once `Impluse` finishes, `close()` of SDF operator will be called, but meanwhile no more processing time timer can be registered. Simply draining timers from operator itself is not ideal. 
Is it possible for us to change something here? For example, the operator should wait for global watermark advancing to MAX_TIMESTAMP to finish? Or the task should invokes `operator.close()` when global watermark advancing to MAX_TIMESTAMP?

Reply:
Commit https://github.com/apache/beam/pull/13105/commits/56a004dd8fd9bb7c52fb56e0f53cd0dc7f8b1dfa introduces a workaround: draining processing timers manually for SDF. It's a workaround since it doesn't work efficiently if upstream advances watermark to MAX_TIMESTAMP very quickly. PTAL, @mxm

Reply:
What you describe happens when you enable checkpointing or when you pass in `--shutdownSourcesAfterIdleMs`. The operator doesn't shutdown then and continues to process timers. The reason we do not enable this by default is that we want to cleanly shutdown pipelines during tests. Once an operator has been shut down, the entire job cannot be checkpointed anymore.

The new workaround looks good to me. We do the same in the `super.close()`method.

Reply:
Thanks for the explanation! It seems like this problem has got attention from Flink: https://issues.apache.org/jira/browse/FLINK-18647.

I'm still feeling confused on when the operator.close() will be called. It seems like it happens under several condition:

- When checkpointing happens

- When source operator closes successfully, the next downstream operator's close() will be called.

Do I understand it correctly?